codex/stores/repostore/store.nim

```diff
@@ -0,0 +1,398 @@
+
+  iter.next = next
+  return success iter
+
+proc createBlockExpirationQuery(maxNumber: int, offset: int): ?!Query =
+  let queryKey = ? createBlockExpirationMetadataQueryKey()
+  success Query.init(queryKey, offset = offset, limit = maxNumber)
+
+method getBlockExpirations*(
+  self: RepoStore,
+  maxNumber: int,
+  offset: int): Future[?!AsyncIter[?BlockExpiration]] {.async, base.} =
+  ## Get iterator with block expirations
+  ##
+
+  without query =? createBlockExpirationQuery(maxNumber, offset), err:
+    trace "Unable to format block expirations query"
+    return failure(err)
+
+  without queryIter =? (await queryT[BlockMetadata](self.metaDs, query)), err:
+    trace "Unable to execute block expirations query"
+    return failure(err)
+
+  let iter = queryIter.map(
+    proc (fut: Future[(?Key, ?!BlockMetadata)]): Future[?BlockExpiration] {.async.} =
+      let (maybeKey, blockMdOrErr) = await fut
+
+      without key =? maybeKey:
+        warn "Entry without a key"
+        return BlockExpiration.none
+
+      without cid =? Cid.init(key.value).mapFailure, err:
+        error "Failed decoding cid", err = err.msg
+        return BlockExpiration.none
+
+      without blockMd =? blockMdOrErr, err:
+        error "Failed fetching metadata for block", cid = cid, err = err.msg
+        return BlockExpiration.none
+
+      BlockExpiration(cid: cid, expiry: blockMd.expiry).some
+  )
+
+  iter.success
+
+method close*(self: RepoStore): Future[void] {.async.} =
+  ## Close the blockstore, cleaning up resources managed by it.
+  ## For some implementations this may be a no-op
+  ##
+
+  trace "Closing repostore"
+
+  if not self.metaDs.isNil:
+    (await self.metaDs.close()).expect("Should meta datastore")
+
+  if not self.repoDs.isNil:
+    (await self.repoDs.close()).expect("Should repo datastore")
+
+###########################################################
+# RepoStore procs
+###########################################################
+
+proc reserve*(self: RepoStore, bytes: NBytes): Future[?!void] {.async.} =
+  ## Reserve bytes
+  ##
+
+  trace "Reserving bytes", bytes
+
+  await self.updateQuotaUsage(plusReserved = bytes)
+
+proc release*(self: RepoStore, bytes: NBytes): Future[?!void] {.async.} =
+  ## Release bytes
+  ##
+
+  trace "Releasing bytes", bytes
+
+  await self.updateQuotaUsage(minusReserved = bytes)
+
+proc start*(self: RepoStore): Future[void] {.async.} =
+  ## Start repo
+  ##
+
+  if self.started:
+    trace "Repo already started"
+    return
+
+  trace "Starting rep"
+  if err =? (await self.updateTotalBlocksCount()).errorOption:
+    raise newException(CodexError, err.msg)
+
+  if err =? (await self.updateQuotaUsage()).errorOption:
+    raise newException(CodexError, err.msg)
+
+  self.started = true
+
+proc stop*(self: RepoStore): Future[void] {.async.} =
+  ## Stop repo
+  ##
+  if not self.started:
+    trace "Repo is not started"
+    return
+
+  trace "Stopping repo"
+  await self.close()
+
+  self.started = false
```

Comment:
nitpick, we usually have new lines at EOF, maybe we should add a github hook to enforce it...